samza-api/src/main/java/org/apache/samza/operators/descriptors/base/system/SystemDescriptor.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.descriptors.base.system;
+
+import com.google.common.base.Preconditions;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.samza.operators.descriptors.base.stream.OutputDescriptor;
+import org.apache.samza.operators.functions.InputTransformer;
+import org.apache.samza.operators.functions.StreamExpander;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.system.SystemStreamMetadata.OffsetType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The base descriptor for a system. Allows setting properties that are common to all systems.
+ *
+ * @param <SubClass> type of the concrete sub-class
+ */
+public abstract class SystemDescriptor<SubClass extends SystemDescriptor<SubClass>> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SystemDescriptor.class);
+  private static final String FACTORY_CONFIG_KEY = "systems.%s.samza.factory";
+  private static final String DEFAULT_STREAM_OFFSET_DEFAULT_CONFIG_KEY = "systems.%s.default.stream.samza.offset.default";
+  private static final String DEFAULT_STREAM_CONFIGS_CONFIG_KEY = "systems.%s.default.stream.%s";
+  private static final String SYSTEM_CONFIGS_CONFIG_KEY = "systems.%s.%s";
+  private static final Pattern ID_PATTERN = Pattern.compile("[\\d\\w-_.]+");
+
+  private final String systemName;
+  private final Optional<String> factoryClassNameOptional;
+  private final Optional<InputTransformer> transformerOptional;
+  private final Optional<StreamExpander> expanderOptional;
```

Comment:
See discussion with Cameron about this above. In general I'm ok with instanceOf checks. But in this case, since both Transforming and Expanding systems can have transform functions, I agree with Cameron that it's simpler to pull these into the base class. Alternatively, we could use an util method to get the transform function from a system descriptor to consolidate the instanceOf checks in one place, but there's still the risk of forgetting to use the method.